external/storm-pmml/src/main/java/org/apache/storm/pmml/runner/jpmml/JPmmlModelRunner.java

```diff
@@ -0,0 +1,134 @@
+        predictedFields = eval.getPredictedFields();
+        outputFields = eval.getOutputFields();
+    }
+
+    /**
+     * @return The raw inputs extracted from the tuple for all 'active fields'
+     */
+    @Override
+    public Map<FieldName, Object> extractRawInputs(Tuple tuple) {
+        LOG.debug("Extracting raw inputs from tuple: = [{}]", tuple);
+        final Map<FieldName, Object> rawInputs = new LinkedHashMap<>();
+        for (FieldName activeField : activeFields) {
+            rawInputs.put(activeField, tuple.getValueByField(activeField.getValue()));
+        }
+        LOG.debug("Raw inputs = [{}]", rawInputs);
+        return rawInputs;
+    }
+
+    @Override
+    public Map<FieldName, FieldValue> preProcessInputs(Map<FieldName, Object> rawInputs) {
+        LOG.debug("Pre processing raw inputs: = [{}]", rawInputs);
+        final Map<FieldName, FieldValue> preProcInputs = new LinkedHashMap<>();
+        for (Map.Entry<FieldName, Object> rawEntry : rawInputs.entrySet()) {
+            preProcInputs.putIfAbsent(rawEntry.getKey(), EvaluatorUtil.prepare(eval, rawEntry.getKey(), rawEntry.getValue()));
+        }
+        LOG.debug("Pre processed inputs = [{}]", preProcInputs);
+        return preProcInputs;
+    }
+
+    @Override
+    public Map<FieldName, ?> predictScores(Map<FieldName, FieldValue> preProcInputs) {
+        LOG.debug("Predicting scores for pre processed inputs: = [{}]", preProcInputs);
+        Map<FieldName, ?> predictedScores = eval.evaluate(preProcInputs);
+        LOG.debug("Predicted scores = [{}]", predictedScores);
+        return predictedScores;
+    }
+
+    /**
+     * @return the predicted scores for the 'predicted fields' and 'output fields'.
+     * All the declared streams will have the same scored tuple.
+     */
+    @Override
+    public Map<Stream, List<Object>> scoredTuplePerStream(Tuple input) {
+        final Map<FieldName, Object> rawInputs = extractRawInputs(input);
+        final Map<FieldName, FieldValue> preProcInputs = preProcessInputs(rawInputs);
+        final Map<FieldName, ?> predScores = predictScores(preProcInputs);
+
+        return toValuesMap(predScores);
+    }
+
+    // Sends the same tuple (list of scored/predicted values) to all the declared streams
+    private Map<Stream, List<Object>> toValuesMap(Map<FieldName, ?> predScores) {
+        final List<Object> scoredVals = new ArrayList<>();
+
+        for (FieldName predictedField : predictedFields) {
+            Object targetValue = predScores.get(predictedField);
+            scoredVals.add(EvaluatorUtil.decode(targetValue));
+        }
+
+        for (FieldName outputField : outputFields) {
+            Object targetValue = predScores.get(outputField);
+            scoredVals.add(EvaluatorUtil.decode(targetValue));
+        }
+
+        final Map<Stream, List<Object>> valuesMap = new HashMap<>();
+
+        for (Map.Entry<? extends Stream, ? extends Fields> entry : modelOutputFields.toMap().entrySet()) {
```

Comment:
use keySet. It may be better if ModelOuputFields itself has a method to returns the keys, right now its more of map wrapper.